README.md
# Zenvidia
This is a bash/zenity script for managing **NVIDIA©** propriatary drivers.

## IMPORTANT NOTICE FOR V1.0
New major update before v1.0.
 - Upgraded to 390.xx new installer options.
 - **.sh** suffixes are removed.
 - Polkit Auth instead of su/sudo.
 - To update do ```sudo make install``` (see **News** below).
 - You will be automaticaly notified for next update. 

## Issues
Nvidia-installer refuse to execute because of nvidia-drm module loaded, the automatic script work-arround does the job.

During the review time, you may meet numbers of issues.

See **Wiki** for **Issues**, **Todo list** and **demands**.

------------
## News

*FIX & UPDATE* (11/28/2018): fix for new 410 serie options sets.
 - remove nvidia-installer driver compil part (definitivly unusable).
 - add per module config install. nvidia uvm and/or drm are optionals and useless for most of us.
 - prime and bumblebleed are still not revised.
 - ISSUE : DKMS is compiling, but not installing modules. Not a clue at this point (Fedora? dkms? nvidia? kernel?). 
 	Anyway, 2nd workaround install drivers after a few fixes.
 	Until debug and final fix of this issue, in case of a kernel upgrade, it is mandatory to use the 'Update driver for an other kernel' menu entry from 'update' menu before restart the computer.

## Notice
I have severe heals issue and I don't really know if I could still manage my project alive. I will try to continue to fix bugs and Nvidia options updates, but i can't tell about the future. It's strongly recommanded to fork.

------------

## Usage

### Driver install
  
 - from local package.
 - from a dowloaded package.
 - from NVIDIA© server.
 - Optimus installation from GIT (bumblebee and prime).
 
### Updates

 - driver updates check.
 - New kernel update (with dkms or not).
 - Optimus GIT sources update.

### Tools

 - Edit xorg.conf file (optimus auto detection).
 - Édit Zenvidia config file.
 - Start Nvidia-Settings (optimus auto detection).
 - Installed driver mangagement (remove, backups).
 - Re-compile some depencies (Bumblebee, etc).

### Tests and support

 - Test GLX.
 - Changelog and driver manual.

## Install
### Zenvidia
This will install in default behaviour with no possible custom ability.

Choose a directory to clone repo and :
```sh
  git clone https://github.com/wildtruc/zenvidia.git
  cd zenvidia/
  # then :
  # to install to default :
  make install
  # to remove all :
  make uninstall
  # to remove safely (doesn't remove downloaded driver packages)
  make safeuninstall
```
And :
```sh
  zenvidia.sh
```
Or by the desktop menu entry in Setting menu.

The script will ask you for admin/superuser password, depend fo which distro you are using.

### Zen Notify
Zenvidia notify is taskbar notifier checking at user session boot time for driver & other GIT repos updates.
It comes with 3 options:
 - -a > check all.
 - -z > check zenvidia script and nvidia drivers.
 - -n > check nvidia drivers only.

Default desktop entry file is set to ```-a```, you can manage options through Zenvidia > Tools.

The script is installed at the same time as Zenvidia when launching ```make install``` command.

## Configuration
Most part of the basic.conf file vars will be updated during the script execution.

You just have to adjust manually :
 - language setting (FR, EN, DE, IT, ES, etc.).
 - default user name: you can leave it at default, but you may change it to your defautl user if you find some issue with user name in your distro behaviour.
 - CUDA setting
 - DKMS setting
 - set Help Tips to 1 or 0, as you like.
 - xterm delay is the time terminal window will stay open.

```
# default desktop user id
def_user=$USER
# locale language for ui
LG=EN
# right/left cairo_dock reserved space, if any (ex: dock='-28')
# this is for centering xterm window correctly.
dock=''

## Basic config vars
# use CUDA module by default (1) or not (0)
cuda=1
# use dkms script by default (1) or not (0)
use_dkms=1
# force use of direct (0) or indirect (1) GL libs
use_indirect=0
# force use of libglvnd (1) or not (0)
use_glvnd=0
# driver install type:  optimus (1), single GPU (0) 
install_type=0
# optimus install type: prime (0), bumblebee (1) 
use_bumblebee=0
# to hold xterm instead of delay: delay (0), hold (1)
xt_hold=0
# extend the delay of xterm (in seconds): default '4'
xt_delay=4
# use help tip: yes (1), no (0)
hlp_txt=1
```

---------

## Licence

Zenvidia is published under GNU/GPL
-----------------------------------

Copyleft PirateProd - Licence GPL v.3

This program is free software; you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation; either version 2.1 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License along with main.c; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor Boston, MA 02110-1301,  USA




CHANGELOG.md
 ## CHANGELOG
 - (2019-10-23) v1.1 downgrade (48)
  - ISSUE: Clean code missing some features and last driver install became wierd. Obviously, I cleaned to much.
  - FIX: Downgraded to previous code until resending a real and functional cleaned code.
  - MEANWHILE: Driver installer add a new dir in /usr/lib, backup/restore was broken because of this. Added also ld.conf.so.d nvidia conf properly to backup/restore function.
  - My apologies for all this. Please backup your driver before doing anything, juste in case.
  
 - (2019-10-01) v1.1 (46)
  - Bug fix: Wrong window height prevent to display "Install from" menu titles.
 
 - (2019-10-01) v1.0.9 (45 last)
  - Final cleaned code. Script is now entirely dedicated to driver, bumblebee and prime are remouved.
  - This is the last time I dive into it. Take care, people.
  
 - (2019-11-08) v1.0.8 (44)
  - fix: Until 415 series, it was needed to replace nvidia's libwfb.so by system xorg ones. Above version don't need it, because libnvidia-wfb doesn't even exist. Then, a filter has been added. (I know, coming a bit late)
  
 - (2019-10-10) v1.0.7 (43)
  - Preparing script for next 'offload' tech for optimus. Cleaning useless code and starting to remove all referenced 'bumblebee' data (project looks dead).
  - There were a bug in extract_build functiion, extractor didn't find the driver directory for an unknow reason. Rewrote function filters expecting to avoid this.
  - Add common vars for install type function, bumblebeed avoided, there no more reason to separate them.
  - Fixed DKMS building disfunction: nutshell in DKMS conf file building (sic!) and replace nvidia-installer option by script's one.
  - Old commented and all bumblebee code will be cleaned next version.
  - Note: Bumblebee project looks dead, but it's not really sure and old device could still have the use of it. Then, I will keep bumblebee code in backup in case of project resurection.
   
 - (2019-08-20) v1.0.6 (42) - urgent fixes (b) and grub update to blsconfig.
  - on fedora 30 and probably many other distros, grub has moved to blscfg, script needed to be update.
  - made urgent fixes on previous updates (dev brain is not so well :s, sorry...)
  - add some missing dependencies in auto install (diff,patch), control is made at each start now.
  - fixed some miss behaviour in librairies linking.
 
 - (2019-08-12) v1.0.5 (41) - fixes and update.
   - zen_notify doesn't warn about not existant repos in Nvidia ftp (rare, but happens). This is fixed.
   - Has been update in zenvidia for the same reason.
   - Added warning display message for incompatible drivers in drivers download window.
   - fixed remote repos download (for old and new drivers).
   - A few minor fixes.
   - Clean code.
   - Bumblebee and prime setup has to be blocked, no enought data to update them at this point (next).
   
 - (2019-08-01) v1.0.4 (40) - fixes on previous update.
   - zen_notify fixes for notify-send displayed text.
 	 	
 - (2019-07-25) v1.0.3 (39) - zen_notify fixes.
   - Changed git local repo "find" command line option to directory only.
   - reverse functions order for a better notification display.
 
 - (2019-07-21) v1.0.2 (38) - changes and fixes.
   - Add a GCC Mismatch option in the case of an error occur in GCC compilator mismatch between driver compiler and kernel compiler. Rare.
   - DKMS part can abort in Fedora environment, add a driver compilation check before starting workaround and then move .ko file to kernel dir if detected.
   - Implement previous dkms driver total remove when upgrade to new driver version. Seems DKMS have issue with that (Fedora env exception ? Nvidia-installer ?). 
   - Restructured nvidia hardware compatibility list.
   - Fixed "back to main" selector in some window that didn't... send back.
   - Old code syntax are still there and commented at this point. Bumblebee and others are still not upgraded.
   
 - (2019-02-02) v1.0.1 (37) - git commmit fix.
   - I've just realize zenvidia pkexec policy rules was not commit correctly. My apology, it shouldn't happen.
 
 - (2018-11-28) v1.0.1 (36) - change and fixes
   - Message (important): I have severe heals issue and I don't really know if I could still manage my project alive. I will try to continue to fix bugs and Nvidia options updates, but i can't tell about the future. It's strongly recommanded to fork.
   - Issue #1: DKMS is compiling, but not install modules. Not a clue at this point (Fedora? dkms? nvidia?). Anyway, 2nd workaround install drivers after a few fix. Until debug and fianl fix of this issue, in case of a kernel upgrade, it is mandatory to use the 'Update driver for an other kernel' entry from 'update' menu before restart the computer.
   - Issue #2: script remove all /opt dir and place it in nvidia's directory when installing a previous version of the driver. Have no clue yet.
   - fix: There was an issue with compiling driver for an other kernel when dkms failed. Had to add env new kernel variables and control to fix.
   - Add a modules chooser at install time : most driver modules are not required for default usage. For a better user experience, a modules selector has been added in case user doesn't use Cuda or Optumus Prime. Use of blacklist was prefered to nvidia-installer options sets. Cuda config option is depracated. No translation pack update for the moment.
   - Update script to new/removed nvidia_installer options 410 serie (libs part).
   - nvidia-installer driver compil part has been remove, because useless (can't build with modules loaded and nvidia-installer have no workaround option, a quite stupid).
   - Not done: remove commented trash codes, bumblebleed/prime sections controls and tests.
   - as usual: please, report if somehing strange happens.
 
 - (2018-03-15) v1.0pre (35) - Some change before code review. 
   - .sh suffixes are removed.
   - zen_notify modified for better design and check.

update.sh
#! /bin/bash

# this script is used both by the Makefile and by zenvidia to update itself.

install_prefix=/usr/local
nvdir=$install_prefix/NVIDIA
nvconf=$install_prefix/etc/zenvidia

## migrate conf files to local/etc (reserve before code review).
#if [[ -s $nvdir/script.conf ]]; then
#	mv -f -t $nvconf $nvdir/{script,basic,color}.conf 
#fi

script_conf=$nvdir/script.conf
basic_conf=$nvdir/basic.conf
color_conf=$nvdir/color.conf
[[ -s $script_conf ]]|| exit 0
. $script_conf
. $color_conf

## create temporary files.
nv_patch=$(mktemp --tmpdir nv_patch-XXXX)
nv_diff=$(mktemp --tmpdir nv_diff-XXXX)
# update conf without remove user's changes.
conf_list=("$script_conf" "$basic_conf" "$color_conf")
for conf in "${conf_list[@]}"; do
	c_old=$conf
	c_new=$(printf "$conf"|sed -n "s/^.*\///p")
	c_orig=$(stat -c "%Y" $c_old)
	c_update=$(stat -c "%Y" ./$c_new)
	if [ $c_update -ne $c_orig ]; then
		diff -u $c_old $c_new &>$nv_patch
		if [[ $conf != $basic_conf ]]; then
			if [ $(cat $nv_patch | grep -c "+") -gt 0 ]; then
				echo -e "$xB# Diff in $c_old:\n>>$xN Applying $c_new diff patch.\n"
				patch -stp0 -i $nv_patch $c_old
			fi
		else
			cat $nv_patch|grep "^[+\|-]\(\w\{1\}\|#\)" &>$nv_diff
			diff_list=$(cat $nv_diff|grep "^[+]\(\w\{1\}\|#\)")
			ifs=$IFS
			IFS=$(echo -en "\n\b")
			for c_list in $diff_list; do
				diff_new=$(printf "$c_list"| grep "+"| sed -n "s/+//p")
				diff_count=$(cat $nv_diff| grep -B 1 "$c_list"| grep -c "-")			
				diff_old=$(cat $nv_diff| grep -B 1 "$c_list"| grep "-"| sed -n "s/-//p")
				if [ $diff_count -eq 0 ]; then
					echo -e "$xB# Diff in $c_old:\n>>$xN $c_list.\n"
					printf "$diff_new\n" >> $c_old
				fi
			done
			IFS=$ifs
		fi
	fi
done
if [ $first_start = 0 ]; then
	sed -i "s/first_start=1/first_start=0/i" $script_conf
fi
## update zenvidia help file
#[ -s ./HELP.md ]&& cp -f ./HELP.md $nvdir/
exit 0


zen_notify.sh
#! /bin/bash

#  Zen_notify
#  mar. oct. 11 16:21:02 CEST 2016
#  Copyright  2010-2016  PirateProd
#  <wildtruc@noneltd.net>
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 2.1 of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with main.c;if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor Boston, MA 02110-1301,  USA

## basics
nvdir=/usr/local/NVIDIA
script_conf=$nvdir/script.conf
basic_conf=$nvdir/basic.conf

################################################
## DEVELOPPEMENT only, DON'T EDIT OR UNCOMMENT'
#devel=/home/mike/Developpement/NVIDIA/zenvidia
#script_conf=$devel/script.conf.devel
################################################

[ $script_conf ]|| exit 0
. $script_conf
. $basic_conf

msg_git="GIT update."
[ -d $nvtmp ]|| mkdir -p $nvtmp
nvtmp=/home/$USER/.zenvidia
ARCH=$(uname -p)
drv_temp=$(mktemp --tmpdir zen_notif-XXXX)
[[ -d $nvtmp ]]|| mkdir -p $nvtmp
local_src=$nvtmp/src
if [[ ! -e home/$USER/.config/autostart/zen_notify.desktop ]]; then
	cp -f $local_src/zenvidia/desktop_files/zen_notify.desktop /home/$USER/.config/autostart/
fi

driver_ctrl(){
#lftp -c "anon; cd ftp://$nvidia_ftp-$ARCH/ ; ls > $nvtmp/drvlist ; cat latest.txt > $nvtmp/last_update"
## get all driver list.
wget -q -O $drv_temp https://$nvidia_ftp-$ARCH/
## get latest.txt file.
wget -q -O $nvtmp/last_update https://$nvidia_ftp-$ARCH/latest.txt
#cat $nvtmp/drvlist_0 |  egrep -o "href.*[0-9]+/'"| perl -pe "s/^.*\'(.*)\/\'/\1/p" > $nvtmp/drvlist
cat $drv_temp| sed -En "s/^.*href.*'(.*)\/'>.*$/\1/p" > $nvtmp/drvlist
cat 	$drv_temp| sed -En "s/^.*href.*'(.*)\/'>.*$/\1/p"| tail -n1 > $nvtmp/last_beta
LAST_DRV=$(cat $nvtmp/last_update | awk '{ print $1 }')
#LAST_BETA=$(cat $nvtmp/drvlist | awk '{ print $9 }' | sort -gr | sed -n 1p)
LAST_BETA=$(cat $nvtmp/last_beta)
#LAST_BETA='370.29'

if [ -s $nvdir/version.txt ]; then
	version=$(cat $nvdir/version.txt| sed -En "s|^(.{6}).*$|\1|;s|\.||p")
else
	version=$(modinfo -F version nvidia| sed -En "s|^(.{6}).*$|\1|;s|\.||p")
fi
offi_short=$(printf "$LAST_DRV"|  sed -En "s|^(.{6}).*$|\1|;s|\.||p")
beta_short=$(printf "$LAST_BETA"| sed -En "s|^(.{6}).*$|\1|;s|\.||p")
[[ $beta_short ]]|| beta_short=0
if [ $beta_short -eq $offi_short ]; then
	DRV_LIST=( "$LAST_DRV,official and beta drivers" )
else
	DRV_LIST=( "$LAST_DRV,official driver" "$LAST_BETA,beta driver" )
fi
[ $version ]|| version=0
if [ ${#DRV_LIST[@]} -gt 0 ]; then
	ifs=$IFS
	IFS=$(echo -en "\n\b")
	for driver in ${DRV_LIST[@]}; do
		drv_short=$(printf "$driver"|cut -d',' -f1 |sed -n "s/\.//p")
		release=$(printf "$driver"|cut -d',' -f2)
		driver=$(printf "$driver"|cut -d',' -f1)
		_driver_notif+=( "$driver $release update.\n" )
	done
	if [ $drv_short -gt $version ]; then
		zenity --notification --window-icon=swiss_knife --text="Nvidia ${_driver_notif[@]}"
	fi
	IFS=$ifs
fi
}
source_ctrl(){
#[[ -d $local_src ]]|| exit 0
for local_git in "${local_git_list[@]}"; do
#	local_git_list=/home/$def_user/tmp/$local_list
	if [[ -d $local_git ]]; then	
		cd $local_git
		git fetch --dry-run &>$nvtmp/git.log	
		if [[ $(cat $nvtmp/git.log|grep -c "master") -eq 1 ]]; then
			local_git_name=$(printf "$local_git"| sed -n "s/^.*\///g;p")
			zenity --notification --window-icon=swiss_knife --text="$local_git_name $msg_git"
		fi
	else
		exit 0
	fi
done
}
usage_help(){
	printf "# Usage : zen_notify option,where option is:\n"
#	printf "\t -a : check all.\n"
	printf "\t -z : check nvidia & zenvidia.\n"
	printf "\t -n : check nvidia only.\n"
	printf "\t -h : this help.\n"
}
while [ $# -gt 0 ]
#	getopts "a-z-n-h" OPT; do
	getopts "z-n-h" OPT; do
	case $OPT in
#		a) driver_ctrl
#		local_src_list=( 'bbswitch' 'Bumblebee' 'primus' 'nvidia-prime-select' 'zenvidia' )
#		source_ctrl
#		;;
		z) driver_ctrl
		local_git_list=( "$HOME/.zenvidia/src/.git" )
		source_ctrl
		;;
		n) driver_ctrl;;
		h) usage_help ;;
	esac
done


zen_notify_standalone.sh
#! /bin/bash

## basics vars
nvdir=/usr/local/NVIDIA
local_src=./zenvidia_src
local_dir='.'
local_tmp=/home/$USER/tmp
nvidia_ftp=download.nvidia.com/XFree86/Linux

# control if desktop file and bitmap exist
if [ ! -e /home/$USER/.config/autostart/zen_notify.desktop ]; then
	cp -f $local_dir/zen_notify.desktop /home/$USER/.config/autostart/
	sed -i "s/Exec=.*$/Exec=zen_notify_standalone.sh/" /home/$USER/.config/autostart/zen_notify.desktop
	[ -e $local_dir/swiss_knife.png ]&& cp -f $local_dir/swiss_knife.png /usr/local/share/pixmaps/
	
fi
# define checker messages
msg_driver="driver update is out"
msg_git="a GIT update is out"
# control if tmp dir exist, then create it if not
[ -d $local_tmp ]|| mkdir -p $local_tmp
# define CPU arch
ARCH=$(uname -p)
# check ftp server for update and form data for script messages
lftp -c "anon; cd ftp://$nvidia_ftp-$ARCH/ ; ls > $local_tmp/drvlist ; cat latest.txt > $local_tmp/last_update"
LAST_DRV=$(cat $local_tmp/last_update | awk '{ print $1 }')
LAST_BETA=$(cat $local_tmp/drvlist | awk '{ print $9 }' | sort -gr | sed -n 1p)
# check current nvidia driver version if any
version=$(modinfo -F version nvidia|sed -n "s/\.//p")
[ $version ]|| version=0
# produce the notification message if ftp check give something
for driver in "$LAST_DRV,official" "$LAST_BETA,beta"; do
	drv_short=$(printf "$driver"|cut -d',' -f1 |sed -n "s/\.//p")
	release=$(printf "$driver"|cut -d',' -f2)
	driver=$(printf "$driver"|cut -d',' -f1)
	if [ $drv_short -gt $version ]; then
		zenity --notification --window-icon=swiss_knife --text="Nvidia $driver $release $msg_driver !"
	fi
done
exit 0


install.sh
#! /bin/bash

### 	WARNING THIS SCRIPT IS FOR SUPERUSER ONLY###

#  zenvidia v0.9
#  Sat Feb  6 16:58:20 2010
#  Copyright  2010  PirateProd
#  <mike@noneltd.net>
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 2.1 of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with main.c;if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor Boston, MA 02110-1301,  USA

[ -e ./script.conf ]|| exit 0
. ./script.conf

end='</span>'
v='<span color=\"#005400\">'		#green
y='<span color=\"#2700FF\">'		#blue
j='<span color=\"#FF6800\">'		#orange
jB='<span color=\"#FF3300\" weight=\"bold\">'
#big red orange title
eN='<span color=\"#FF3300\" weight=\"bold\" font=\"20\">'
#Bold
vB='<span color=\"#005400\" weight=\"bold\">'

user_id(){
	if [[ $(whoami) != "root" ]] ; then
#		pass=$(zenity --password --text="$v\Enter SuperUser password$end")
		if [ -f /proc/version ] ; then
			distro_list=$(echo -e "ubuntu\ndebian\nfedora\nredhat\nmandriva")
			for distro in $distro_list ; do
				proc_version=$(cat /proc/version | grep $distro)
				if [[ $proc_version != '' ]] ; then
					proc_version=$distro
					if [[ $(echo $proc_version | grep "ubuntu\|debian") != '' ]] ; then
						SU="sudo"
					else
						SU="su -c"
					fi
				fi
			done
			zenity --password --text="$v\Enter SuperUser password$end"| $SU "sh $0"
			exit 0
		else
			zenity --width=450 --error \
			--text="$v SORRY, CAN'T IDENTIFY DISTRO.\nPROMPT DIRECTLY AS SU\nAND TYPE $J sudo $(basename $0)$end$v FOR DEBIAN LIKE,\nOR$j su -c $(basename $0)$end$v FOR OTHER DISTRO.$end"
			exit 0
		fi
	fi
}
menu(){
#zenity --width=450 --title="Zenvidia" --question \
menu_box=$(zenity --height=460 --title="Zenvidia" --list --radiolist --hide-header \
	--ok-label="Continue" --cancel-label="Quit" \
	--text="$eN\Zenvidia Installer$end\n
$vB\You're going to install Zenvidia and Bashvidia on your system$end
$v\nInstaller will ask you some question:
\t- Default system user name.
\t- Where to install Zenvidia files.
\t- Where to install drivers files.
\t- Where to install Nvidia binary files.
\nYou can leave set to default or change them as appropriate.
$jB\Don't edit the script.conf file before install, it is useless.$end
\nClick$end $j\Continue$end$v button to install.$end" \
	--column "1" --column "2" --column "3" --separator=";" --hide-column=2 \
	true 1 "Install" false 2 "Edit config" false 3 "About scripts") 
if [ $? = 1 ]; then exit 0; fi
	case $menu_box in
		"1") fn_install ;;
		"2") fn_edit_conf ;;
		"3") fn_readme ;;
	esac
}
fn_install(){
for pre_entry in {$USER,$nvdir,$croot,$install_dir,$LG}; do
	if [ $pre_entry = $USER ]; then ENTRY="Default User"
	elif [ $pre_entry = $nvdir ]; then ENTRY="Zenvidia directory"
	elif [ $pre_entry = $croot ]; then ENTRY="Drivers directory"
	elif [ $pre_entry = $install_dir ]; then ENTRY="Nvidia binaries directory"
	elif [ $pre_entry = $LG ]; then ENTRY="Language pack (FR,EN)";fi
	conf+=$(zenity --width=300 --title="Zenvidia" --entry \
	--entry-text="$pre_entry" --text="$ENTRY"),
	if [ $? = 1 ]; then exit 0; fi
done

conf=$(echo "$conf"|sed -n 's/\//\\\//g;s/,$//p')

default_user=$(printf "$conf"|cut -d"," -f1)
default_nvdir=$(printf "$conf"|cut -d"," -f2)
default_croot=$(printf "$conf"|cut -d"," -f3)
default_install_dir=$(printf "$conf"|cut -d"," -f4)
default_LG=$(printf "$conf"|cut -d"," -f5)

sed -ni "s/def_user=\(.*\)/def_user=$default_user/i;p" ./script.conf
sed -ni "s/nvdir=\(.*\)/nvdir=$default_nvdir/i;p" ./script.conf
sed -ni "s/croot=\(.*\)/croot=$default_croot/i;p" ./script.conf
sed -ni "s/install_dir=\(.*\)/install_dir=$default_install_dir/i;p" ./script.conf
sed -ni "s/LG=\(.*\)/LG=$default_LG/i;p" ./script.conf

if [[ $(cat ./script.conf|grep "def_user=$default_user") != '' ]]; then
	. ./script.conf
fi

for binary in {"bashvidia.sh","zenvidia.sh",nvidia-installer}; do
#	echo "copy $binary to $install_dir/bin/"
	cp -f $binary $install_dir/bin/
done
mkdir -p $nvdir/locale $croot
for files in {"script.conf",FR_PACK,EN_PACK,README.md}; do
#	echo "copy $files to $nvdir"
	if [ $(echo "$files"| grep "PACK") != '' ]; then
		cp -f $files $nvdir/locale/
	fi
	cp -f $files $nvdir/ 
done
}
# control script conf
fn_edit_conf(){
	edit_script=$(zenity --width=500 --height=400 --title="Zenvidia" --text-info \
	--editable --text="$v\Edit script config file$end" --filename="$nvdir/script.conf" \
	--checkbox="Confirm to overwrite" )
	if [[ $(printf "$edit_script"| sed -n '1p') != '' ]]; then
		printf "$edit_script" > $script_conf
	fi
	menu
}
fn_readme(){
	zenity --width=500 --height=400 --title="Zenvidia" --text-info \
	--text="$v\About Zenvidia/Bashvidia$end" --filename="./README.md"
	menu
}
user_id; menu
exit 0
